Add VMStep.OverAsyncSuspension and atAsyncSuspension

DanTup opened the pull request:
This fixes #30 and #31.

I used the `Over` test as a template and just made it async; let me know if I can do better. I wasn't totally sure what this code was doing in the `Over` test:

```
stdout.next.then(neverCalled).catchError((_) {});
```

So I didn't include it. It seems like it'll throw on the next output, but then swallow the error; but that seems pointless. LMK what it's actually doing and whether I should have something similar.

I had to do this work against a slightly older SDK (I used v2 dev40) because the tests all time out in the newest versions and I couldn't totally figure out why. I can't see any obvious CI set up here so if it's building somewhere and this fails (I don't expect it to, it passes locally) let me know.

DanTup committed:
Add support for OverAsyncSuspension

Fixes #30
Fixes #31

jacob314 commented on lib/src/pause_event.dart:
```diff
@@ -25,6 +25,13 @@ VMPauseEvent newVMPauseEvent(Scope scope, Map json) {
 
 /// An event indicating that an isolate has been paused or resumed.
 abstract class VMPauseEvent {
+  /// Is the isolate paused at an await, yield, or yield* statement?
```
Nit: 
Tweak comment to 
/// Whether the isolate is paused at an await, yield, or yield* statement.

side note: a yield statement is an async suspension? I thought that was sync although I haven't used yield in my code.

DanTup replied:
I'll update the text tomorrow. The text was taken from [here](https://github.com/dart-lang/sdk/blob/master/runtime/vm/service/service.md#event) and includes yield. I don't know if the naming is accurate, but I guess the expectation is the same when a user hits "step over" on a yield - it should pause on the next statement in their code, even though though there was other execution in between.

jacob314 commented on test/isolate_test.dart:
```diff
@@ -430,6 +430,51 @@ void main() {
+          await inner(); // line 9
+          print("after inner"); // line 10
+        }
+      """, main: r"""
```
trailing comma style might make this code more readable.
E.g. something like
```dart
 await runAndConnect(
    topLevel: r"""
        inner() {
          print("in inner");
          return new Future.delayed(const Duration(milliseconds: 1));
        }

        outer() async {
          debugger();
          await inner(); // line 9
          print("after inner"); // line 10
        }
      """,
    main: r"""
        await outer();
        print("after outer");
      """,
  )   


trailing comma style might make this code more readable.

DanTup replied:
The only difference it makes is that the closing paren goes on the next line. This is formatted the same as the rest of the file though, so I'm not sure if to commit that.

jacob314 reviewed (approved):
Please wait for @nex3 to approve as well.

jacob314 commented:
Please also update the changelog and modify the version number. It would be fine to only perform one version update for both this CL and your other pending CL to vm_service_client.

DanTup committed:
Update changelog and minor comments

DanTup commented:
Addressed the comments and updated changelog/version.

I'll update the version to the same in the other PR and rebase whichever PR comes last and merge together (I didn't want to add the other PRs changes to the changelog here as that seemed a bit weird).

DanTup committed:
Add note about VM support

DanTup commented:
(btw, someone will need to merge this when happy with it as I don't have merge permissions here)

DanTup committed:
Reformat file

DanTup commented:
I've rebased on master since the version number I used has been used now and also fixed a minor formatting issue that made Travis red.

jacob314 commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+## 0.2.5+2
```
@nex3 had to point it out to me as well but this should technically be
0.2.6 instead of 0.2.5+2
as you added functionality instead of just making bug fixes.
https://semver.org/

jacob314 reviewed (commented):
approved to land after the version is tweaked to 0.2.6 and the pubspec.yaml is updated as well.

DanTup committed:
Bump version to 0.2.6
